text/0000-canonical-metadata.md

```diff
@@ -0,0 +1,1232 @@
+        * type
+        * options: composed, bubbles
+    * Programmatically fetched static resources
+        * `<script>`
+        * `<img>`
+        * `<link>`
+* CSS file
+    * Tokens
+    * Static resources
+
+## Metadata shape
+
+### TypeScript
+
+#### Bundle metadata shape
+```ts
+interface BundleMetadata {
+    version: string;
+    name?: string;
+    namespace?: string;
+    // Define the type of module that is represented by the given metadata object.
+    moduleType?: "HTML" | "CSS" | "Javascript";
+
+    /**
+     * For a component bundle, this property is a reference to the main component class in the bundle.
+     * This class must be the default export of the javascript file. The file must be named the same
+     * as the bundle
+     */
+    defaultComponentClass?: Class;
+
+    // Information about HTML template files in the bundle.
+    templates?: HTMLTemplateFile[];
+
+    // Information about javascript files in the bundle.
+    scripts?: ScriptFile[];
+
+    // Information about css files in the bundle.
+    css?: CSSFile[]
+}
+```
+Other alternatives:
+- Indexed by file type and file name
+```ts
+interface BundleMetadataOption2 {
+    version: string;
+    name?: string;
+    namespace?: string;
+    moduleType?: "HTML" | "CSS" | "Javascript";
+
+    defaultComponentClass?: ComponentClass[];
+
+    templates?: { [key: string]: HTMLTemplateFile[]};
+
+    scripts?: { [key: string]: ScriptFile[]};
+
+    css?: { [key: string]: CSSFile[]};
+}
+```
+
+#### HTML Template file metadata shape
+```ts
+// Root metadata object for an HTML template file
+interface HTMLTemplateFile {
+    fileType: 'html';
+    fileName: string;
+    // List of unique components referenced in the template
+    componentReferences?: ComponentReference[];
+    // Static resources referenced in a template. For example: 'src' attribute value of an <img>, <source> tag.
+    staticResources?: StaticResourceReference[];
+    // Template in AST format starting with the root <template> node. Note: this is a partial AST
+    astRoot?: Node;
+}
+
+// Represents a component referenced in the template.
+// Use the 'tagName' property for kebab case and 'name' property for the reference in canonical form
+interface ComponentReference extends ModuleReference {
+    // Component reference always has a namespace
+    namespace: string;
+    // No moduleIdentifier for a component
+    moduleIdentifier?: never;
+    // Components are always imported from an external bundle
+    type: 'external';
+    tagName: string;
+}
+
+// Information about a node in the html template
+interface Node {
+    tagName: string;
+    location: SourceLocation;
+    // Any properties set in the template, will include html attributes, public properties
+    attributes?: ElementAttributeReference[];
+    // Event listeners attached declaratively in the template with a 'on' prefix, eg: 'onclick'
+    eventListeners?: DeclarativeEventListener[];
+    children?: Node[];
+}
```

Comment:
depending on the AST implementation, we may want to have a reference to the parent node.